runtime/obj.c

```diff
@@ -300,3 +300,81 @@ struct queue_chunk {
   struct queue_chunk *next;
   value entries[ENTRIES_PER_QUEUE_CHUNK];
 };
+
+
+/* For compiling let rec over values */
+
+/* [size] is a [value] representing number of words (fields) */
+CAMLprim value caml_alloc_dummy(value size)
+{
+  mlsize_t wosize = Long_val(size);
+  return caml_alloc (wosize, 0);
+}
+
+/* [size] is a [value] representing number of floats. */
+CAMLprim value caml_alloc_dummy_float (value size)
+{
+  mlsize_t wosize = Long_val(size) * Double_wosize;
+  return caml_alloc (wosize, 0);
+}
+
+CAMLprim value caml_update_dummy(value dummy, value newval)
+{
+  mlsize_t size;
+  tag_t tag;
+
+  tag = Tag_val (newval);
+
+  if (Wosize_val(dummy) == 0) {
+      /* Size-0 blocks are statically-allocated atoms. We cannot
+         mutate them, but there is no need:
+         - All atoms used in the runtime to represent OCaml values
+           have tag 0 --- including empty flat float arrays, or other
+           types that use a non-0 tag for non-atom blocks.
+         - The dummy was already created with tag 0.
+         So doing nothing suffices. */
+      CAMLassert(Wosize_val(newval) == 0);
+      CAMLassert(Tag_val(dummy) == Tag_val(newval));
+  } else if (tag == Double_array_tag){
+    CAMLassert (Wosize_val(newval) == Wosize_val(dummy));
+    CAMLassert (Tag_val(dummy) != Infix_tag);
+    Unsafe_store_tag_val(dummy, Double_array_tag);
+    size = Wosize_val (newval) / Double_wosize;
+    for (mlsize_t i = 0; i < size; i++) {
+      Store_double_flat_field (dummy, i, Double_flat_field (newval, i));
+    }
+  } else {
+    CAMLassert (tag < No_scan_tag);
+    CAMLassert (Tag_val(dummy) != Infix_tag);
+    Unsafe_store_tag_val(dummy, tag);
+    size = Wosize_val(newval);
+    CAMLassert (size == Wosize_val(dummy));
+    /* See comment above why this is safe even if [tag == Closure_tag]
+       and some of the "values" being copied are actually code pointers. */
+    for (mlsize_t i = 0; i < size; i++){
+      caml_modify (&Field(dummy, i), Field(newval, i));
+    }
+  }
+  return Val_unit;
+}
+
+CAMLprim value caml_update_dummy_lazy(value dummy, value newval)
+{
+  // Note: [obj_tag] works on immediates as well
+  int tag = obj_tag (newval);
+  switch (tag) {
+  case Lazy_tag:
+  case Forcing_tag:
+  case Forward_tag:
+    caml_update_dummy(dummy, newval);
+    break;
+  // If the tag of [newval] is not a lazy tag,
+  // it comes from a Forward block that was shortcut.
+  default:
+    CAMLassert (Wosize_val(dummy) == 1);
+    caml_modify(&Field(dummy, 0), newval);
+    Unsafe_store_tag_val(dummy, Forward_tag);
```

Comment:
It probably doesn't matter, but `caml_update_dummy` writes the tag before the contents, so I would have done the same here for consistency.

Reply:
I hesitated, but I don't want to have to worry about an intermediate state where the runtime sees `Forward_tag` with a type-invalid payload.